### ISSUE TITLE
No equivalent to theano.tensor.slinalg.Eigvalsh

### ISSUE THREAD

**Issue by tchaton** [tensorflow/tensorflow#18083]
Good morning,

I have been trying to reproduce DeepLDA loss function implementation in Tensorflow 
(https://github.com/VahidooX/DeepLDA/blob/master/code/objectives.py theano version).
In order to solve a gep, in theano we use T.slinalg.eigvalsh to compute eigenvalues
evals_t = T.slinalg.eigvalsh(Sb_t, St_t)

Tensorflow doesn't provide something similar and approximation are very unstable.
I tried both to solve the gep Aei = Bviei
cho = tf.cholesky(B_t + tf.eye(dim) * r ) 
inv_cho = tf.matrix_inverse(cho)
evals_t = tf.linalg.eigvalsh(inv_cho * A * tf.transpose(inv_cho)) 

evals_t =tf.linalg.eigvalsh(tf.matrix_inverse(B) * A )
If you have a solution, please tell me :)

**Comment by tchaton** [comment]
Ubuntu 14.6
Tensorflow==1.4.0, gpu version.
Cuda 8.0
Titan X
I will add the code later on for you to try it out. It works fine using theano implemented LDA loss and backend. I might miss some information. I will go everything next week.

**Comment by tchaton** [comment]
http://www.netlib.org/lapack/lug/node54.html GEP with cholesky
import tensorflow as tf


def lda_loss(n_components, margin):
    """
    The main loss function (inner_lda_objective) is wrapped in this function due to
    the constraints imposed by Keras on objective functions
    """
    def inner_lda_objective(y_true, y_pred):
        """
        It is the loss function of LDA as introduced in the original paper. 
        It is adopted from the the original implementation in the following link:
        https://github.com/CPJKU/deep_lda
        Note: it is implemented by Theano tensor operations, and does not work on Tensorflow backend
        """
        r = 1e-4

        # init groups
        #yt = tf.cast(tf.contrib.layers.flatten(y_true), tf.float32)
        #indexes = tf.argmax (y_true, axis=-1)
        locations = tf.where (tf.equal (y_true, 1))
        indices = locations[:, 1]
        y, idx = tf.unique(indices)
        
        def fn(unique, indexes, preds):
            u_indexes = tf.where(tf.equal(unique, indexes))
            u_indexes = tf.reshape(u_indexes, (1, -1))
            X = tf.gather(preds, u_indexes)
            X_mean = X - tf.reduce_mean(X, axis=0)
            m = tf.cast(tf.shape(X_mean)[1], tf.float32)
            #X_mean = tf.squeeze(X_mean)
            return (1/(m-1)) * tf.matmul(tf.transpose(X_mean[0]), X_mean[0])

        # scan over groups
        covs_t = tf.map_fn(lambda x: fn(x, indices, y_pred), y, dtype=tf.float32)

        # compute average covariance matrix (within scatter)
        Sw_t = tf.reduce_mean(covs_t, axis=0)
        
        # compute total scatter
        Xt_bar = y_pred - tf.reduce_mean(y_pred, axis=0)
        m = tf.cast(tf.shape(Xt_bar)[1], tf.float32)
        St_t = (1/(m-1)) * tf.matmul(tf.transpose(Xt_bar), Xt_bar)

        # compute between scatter
        dim = tf.shape(y)[0]
        Sb_t = St_t - Sw_t

        # cope for numerical instability (regularize)
        
        Sw_t += tf.eye(dim) * r    
        
        cho = tf.cholesky(St_t + tf.eye(dim) * r ) 
        inv_cho = tf.matrix_inverse(cho)
        evals_t = tf.linalg.eigvalsh(inv_cho * Sb_t * tf.transpose(inv_cho)) # Sb_t, St_t    
        #evals_t = tf.abs(tf.linalg.eigvalsh(tf.matrix_inverse(Sw_t) * Sb_t ))
        top_k_evals = evals_t[-n_components:]
        
        #index_max = tf.argmax(top_k_evals, 0)
        #thresh_max = top_k_evals[index_max] - margin
        
        index_min = tf.argmin(top_k_evals, 0)
        thresh_min = top_k_evals[index_min] + margin
        #thresh = tf.contrib.distributions.percentile(top_k_evals, 33)
        #mask_max = top_k_evals > thresh_max
        mask_min = top_k_evals < thresh_min
        
        #cost_max = tf.boolean_mask(top_k_evals, mask_max) 
        cost_min = tf.boolean_mask(top_k_evals, mask_min)
        
        return  - tf.reduce_mean(cost_min)
    
    return inner_lda_objective

import pickle
import gzip
import numpy as np
from keras.datasets import mnist
from keras.optimizers import Adam


if __name__ == '__main__':
    ############
    # Parameters Section
    from tensorflow.contrib.learn.python.learn.datasets.mnist import read_data_sets
    mnist = read_data_sets("path_mnist/", one_hot=True)
    # the path to save the final learned features


    # the size of the new space learned by the model (number of the new features)
    outdim_size = 10

    # the parameters for training the network
    epoch_num = 100
    batch_size = 800

    # the regularization parameter of the network
    reg_par = 1e-5

    # The margin and n_components (number of components) parameter used in the loss function
    # n_components should be at most class_size-1
    margin = 1.0
    n_components = 9

    # Parameter C of SVM
    C = 1e-1
    # end of parameters section
    ############

    # Load data
    (x_train, y_train), (x_test, y_test) = mnist.load_data()

    x_train = np.reshape(x_train, (len(x_train), -1))
    x_test = np.reshape(x_test, (len(x_test), -1))

    print(x_train.shape, y_train.shape)
    # Building, training, and producing the new features by Deep LDA
    model = create_model(x_train.shape[-1], reg_par, outdim_size)

    model_optimizer = Adam()
    model.compile(loss=lda_loss1(n_components, margin), optimizer=model_optimizer)

    model.summary()

    model.fit(x_train, y_train, batch_size=batch_size, epochs=epoch_num, shuffle=True, validation_data=(x_test, y_test), verbose=2)

**Comment by tensorflowbutler** [comment]
Nagging Assignee @skye: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by skye** [comment]
@fchollet can you comment or redirect? Thanks.

**Comment by tchaton** [comment]
It is still not possible to sovle a gpe in tf where it is wth theano. I
tried over way, but tf.solve_cholesky is very unstable.

2018-04-25 2:14 GMT+01:00 Skye Wanderman-Milne <notifications@github.com>:

**Comment by tensorflowbutler** [comment]
Nagging Assignee @fchollet: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @fchollet: It has been 31 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @fchollet: It has been 46 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @fchollet: It has been 61 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @fchollet: It has been 76 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @fchollet: It has been 91 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @fchollet: It has been 105 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @fchollet: It has been 136 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @fchollet: It has been 151 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @fchollet: It has been 165 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @fchollet: It has been 180 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @fchollet: It has been 196 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @fchollet: It has been 211 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by sushreebarsa** [comment]
@tchaton Could you try this alternative `tf.linalg.eigvalsh`  function which closely resembles Eigvalsh. It directly calculates the eigenvalues of a Hermitian positive definite matrix. Please refer the following example;
```
import tensorflow as tf

# Create your Hermitian positive definite matrix (example)
matrix = tf.constant([[2., 1.], [1., 3.]])

# Calculate eigenvalues
eigenvalues = tf.linalg.eigvalsh(matrix)

# Access the eigenvalues (tf.Tensor)
print(eigenvalues)


```

You are using an older version of TF which is not actively supported so kindly use the latest TF version and raise a new ticket if the issue still persists?
Thank you!
